### ISSUE TITLE
support for Git LFS

### ISSUE THREAD

**Issue by jeras** [cdsgit/cdsgit#12]
https://git-lfs.github.com/
Using LFS for binary files would speed up cloning.
And many design files are binary.

Support for locking would also be appreciated, but it is not yet universally supported:
https://jira.atlassian.com/browse/BSERV-9623
https://gitlab.com/gitlab-org/gitlab-ee/issues/1830

**Comment by cookacounty** [comment]
Git LFS is on my roadmap #9 . I am still working on getting an LFS server setup to test this functionality.

As far as locking goes, I don't see this happening on my end in the near future. Another thought I had with the locking would be to use git-svn, but I haven't played with it enough to understand the limitations.

**Comment by cookacounty** [comment]
Do you envision having all the cadence database *.oa files managed via git-lfs?

**Comment by jeras** [comment]
I am new to Virtuoso, before I only used Incisive, and Git in general for C, Python, Verilog, Xilinx FPGA projects, ... so I am not sure yet what to do with open access files.

Actually I did not use Git LFS yet, I did not have too. But Virtuoso generates many binary files, and LTS should provide an advantage. I should start experimenting soon, and hopefully should be able to provide a patch (pull request).

Locking should be implemented in Git-LFS, at least according to change logs. I will try to test it in the next weeks and give you some impressions.

**Comment by jeras** [comment]
Today I tried to use LFS on Bitbucket and GitHub, LFS itself worked in both cases.

There are some possible issued with LFS in combination with repo moves an forks.
GitHub issues are related to who is going to pay for the bandwidth.

Bitbucket provides two conflicting documents regarding LFS repo forking:
https://confluence.atlassian.com/bitbucket/current-limitations-for-git-lfs-with-bitbucket-828781638.html#CurrentlimitationsforGitLFSwithBitbucket-GitLFSfilesarenotcopiedwhenarepoisforked
https://confluence.atlassian.com/bitbucketserver/git-large-file-storage-794364846.html#GitLargeFileStorage-RepositoryForking

I also tried locking. Bitbucket did not accept locking requests, this was expected, see Bitbucket and Gitlab issues linked above.
https://github.com/git-lfs/git-lfs/wiki/File-Locking
With GitHub locking worked as documented, but this is simple single branch locking, not the advanced locking described in the original locking proposal.
I have seen mentioned that a Microsoft git cloud platform supports locking.

Next I will try to run my own LFS server. I intend to try this one:
https://github.com/artemkin/git-lfs-server

**Comment by jeras** [comment]
Hi @cookacounty,

I started evaluating https://gvfs.io/ from MS. It should offer an even better performance while keeping all the files in the same git repo instead of storing binary data in a separate storage (which means easier migrations).
I found the more interesting part is the history of the project, since it provides a comparison to other solutions.
https://www.visualstudio.com/learn/gvfs-design-history/

Unfortunately a Linux implementation is still far away. But I do not have a very large repo yet, so I do not need a binary file solution yet.

Locking would still be useful, but I did not see it mentioned in combination with GVFS yet.
